File: behavioral/typescript/visitor_structured/interfaces.ts
export interface Visitor {
  visitBook(book: any): void;
  visitFruit(fruit: any): void;
  visitElectronic(electronic: any): void;
}

export interface Item {
  accept(visitor: Visitor): void;
  getPrice(): number;
}

export interface FileSystemVisitor {
  visitFile(file: any): void;
  visitDirectory(directory: any): void;
}

export interface FileSystemNode {
  accept(visitor: FileSystemVisitor): void;
  getName(): string;
  getSize(): number;
}

export interface SalaryVisitor {
  visitEmployee(employee: any): void;
  visitManager(manager: any): void;
  visitDeveloper(developer: any): void;
  visitDesigner(designer: any): void;
}

export interface Employee {
  accept(visitor: SalaryVisitor): void;
  getName(): string;
  getBaseSalary(): number;
}

File: behavioral/dart/visitor_structured/interfaces.dart
abstract class Visitor {
  void visitBook(Book book);
  void visitFruit(Fruit fruit);
  void visitElectronic(Electronic electronic);
}

abstract class Item {
  void accept(Visitor visitor);
  double getPrice();
}

abstract class FileSystemVisitor {
  void visitFile(File file);
  void visitDirectory(Directory directory);
}

abstract class FileSystemNode {
  void accept(FileSystemVisitor visitor);
  String getName();
  int getSize();
}

abstract class SalaryVisitor {
  void visitEmployee(Employee employee);
  void visitManager(Manager manager);
  void visitDeveloper(Developer developer);
  void visitDesigner(Designer designer);
}

abstract class Employee {
  void accept(SalaryVisitor visitor);
  String getName();
  double getBaseSalary();
}
